app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/yeartodatelliability/YearToDateLiabilityController.scala

```diff
@@ -0,0 +1,1110 @@
+                            val nowRequiresPersonalAllowance = i > 0
+
+                            if (hadRequiredPersonalAllowance =!= nowRequiresPersonalAllowance)
+                              incomplete.copy(estimatedIncome = Some(estimatedIncome), personalAllowance = None)
+                            else
+                              incomplete.copy(estimatedIncome = Some(estimatedIncome))
+                          },
+                          _ =>
+                            IncompleteCalculatedYearToDateLiabilityAnswers.empty
+                              .copy(estimatedIncome = Some(estimatedIncome))
+                        )
+
+                      draftReturn.copy(yearToDateLiabilityAnswers = Some(newAnswers))
+                    }
+                }
+              )
+            }
+
+          case _ =>
+            Redirect(routes.YearToDateLiabilityController.checkYourAnswers())
+        }
+    }
+  }
+
+  def personalAllowance(): Action[AnyContent] = authenticatedActionWithSessionData.async { implicit request =>
+    withFillingOutReturnAndYTDLiabilityAnswers(request) {
+      case (_, fillingOutReturn, answers) =>
+        (answers, fillingOutReturn.draftReturn) match {
+          case (calculatedAnswers: CalculatedYearToDateLiabilityAnswers, draftReturn: SingleDisposalDraftReturn) =>
+            withDisposalDate(draftReturn) { disposalDate =>
+              withEstimatedIncome(calculatedAnswers) { estimatedIncome =>
+                if (estimatedIncome.value > 0L) {
+                  commonDisplayBehaviour(calculatedAnswers)(
+                    form = { a =>
+                      val emptyForm = personalAllowanceForm(disposalDate)
+                      a.fold(
+                        _.personalAllowance.fold(emptyForm)(a => emptyForm.fill(a.inPounds())),
+                        _.personalAllowance.fold(emptyForm)(a => emptyForm.fill(a.inPounds()))
+                      )
+                    }
+                  )(
+                    page = personalAllowancePage(_, _, disposalDate)
+                  )(
+                    requiredPreviousAnswer = _.fold(_.estimatedIncome, c => Some(c.estimatedIncome)),
+                    routes.YearToDateLiabilityController.estimatedIncome()
+                  )
+                } else {
+                  Redirect(routes.YearToDateLiabilityController.checkYourAnswers())
+                }
+              }
+            }
+
+          case _ =>
+            Redirect(routes.YearToDateLiabilityController.checkYourAnswers())
+        }
+    }
+  }
+
+  def personalAllowanceSubmit(): Action[AnyContent] = authenticatedActionWithSessionData.async { implicit request =>
+    withFillingOutReturnAndYTDLiabilityAnswers(request) {
+      case (_, fillingOutReturn, answers) =>
+        (answers, fillingOutReturn.draftReturn) match {
+          case (calculatedAnswers: CalculatedYearToDateLiabilityAnswers, draftReturn: SingleDisposalDraftReturn) =>
```

Comment:
this pattern matching is new - some pages are only shown for a calculated journey  for single disposals